src/storage/engine/mod.rs

```diff
@@ -81,96 +81,233 @@ pub trait Snapshot: Send {
+    fn iter<'a>(&'a self, upper_bound: Option<&[u8]>, mode: ScanMode) -> Result<Cursor<'a>>;
     #[allow(needless_lifetimes)]
-    fn iter_cf<'a>(&'a self, cf: CfName, upper_bound: Option<&[u8]>) -> Result<Box<Cursor + 'a>>;
+    fn iter_cf<'a>(&'a self,
+                   cf: CfName,
+                   upper_bound: Option<&[u8]>,
+                   mode: ScanMode)
+                   -> Result<Cursor<'a>>;
 }
 
-pub trait Cursor {
+pub trait Iterator {
     fn next(&mut self) -> bool;
     fn prev(&mut self) -> bool;
     fn seek(&mut self, key: &Key) -> Result<bool>;
     fn seek_to_first(&mut self) -> bool;
     fn seek_to_last(&mut self) -> bool;
     fn valid(&self) -> bool;
 
+    fn validate_key(&self, _: &Key) -> Result<()> {
+        Ok(())
+    }
+
     fn key(&self) -> &[u8];
     fn value(&self) -> &[u8];
+}
+
+macro_rules! near_loop {
+    ($cond:expr, $fallback:expr) => ({
+        let mut cnt = 0;
+        while $cond {
+            cnt += 1;
+            if cnt >= SEEK_BOUND {
+                CURSOR_OVER_SEEK_BOUND_COUNTER.inc();
+                return $fallback;
+            }
+        }
+    })
+}
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum ScanMode {
+    Forward,
+    // No Backward here, because we won't just use pure Backward.
+    Mixed,
+}
+
+pub struct Cursor<'a> {
+    iter: Box<Iterator + 'a>,
+    scan_mode: ScanMode,
+    // the data cursor can be seen will be
+    min_key: Option<Vec<u8>>,
+    max_key: Option<Vec<u8>>,
+}
+
+impl<'a> Cursor<'a> {
+    pub fn new<T: Iterator + 'a>(iter: T, mode: ScanMode) -> Cursor<'a> {
+        Cursor {
+            iter: Box::new(iter),
+            scan_mode: mode,
+            min_key: None,
+            max_key: None,
+        }
+    }
+
+    pub fn seek(&mut self, key: &Key) -> Result<bool> {
+        if self.max_key.as_ref().map_or(false, |k| k <= key.encoded()) {
+            try!(self.iter.validate_key(key));
+            return Ok(false);
+        }
+
+        if self.scan_mode == ScanMode::Forward && self.valid() && self.iter.key() >= key.encoded() {
+            return Ok(true);
+        }
+
+        if !try!(self.iter.seek(key)) {
+            self.max_key = Some(key.encoded().to_owned());
+            return Ok(false);
+        }
+        Ok(true)
+    }
 
     /// Seek the specified key.
     ///
     /// This method assume the current position of cursor is
     /// around `key`, otherwise you should use `seek` instead.
-    fn near_seek(&mut self, key: &Key) -> Result<bool> {
-        if !self.valid() {
+    pub fn near_seek(&mut self, key: &Key) -> Result<bool> {
+        if !self.iter.valid() {
             return self.seek(key);
         }
-        if self.key() == &**key.encoded() {
+        let ord = self.iter.key().cmp(key.encoded());
+        if ord == Ordering::Equal ||
+           (self.scan_mode == ScanMode::Forward && ord == Ordering::Greater) {
             return Ok(true);
```

Comment:
No, because the cursor is always sought forward, so `iter.prev() < key` must be true..